File: GameLibrary/test/cc/lib/game/TestUtils.java
package cc.lib.game;

import junit.framework.TestCase;

import org.junit.Assert;

import java.awt.Rectangle;
import java.util.Arrays;
import java.util.List;

public class TestUtils extends TestCase {

    public void test_isBoxesOverlapping() {
        Rectangle A = new Rectangle(0,0,1,1);
        Rectangle B = new Rectangle(-1,-1,3,3);
        Rectangle C = new Rectangle(-2, 0, 5, 1);
        
        Rectangle r0 = A;
        Rectangle r1 = B;       
        assertTrue(Utils.isBoxesOverlapping(r0.x, r0.y, r0.width, r0.height, r1.x, r1.y, r1.width, r1.height));
        assertTrue(Utils.isBoxesOverlapping(r1.x, r1.y, r1.width, r1.height, r0.x, r0.y, r0.width, r0.height));
        
        r0 = A;
        r1 = A;
        assertTrue(Utils.isBoxesOverlapping(r0.x, r0.y, r0.width, r0.height, r1.x, r1.y, r1.width, r1.height));
        assertTrue(Utils.isBoxesOverlapping(r1.x, r1.y, r1.width, r1.height, r0.x, r0.y, r0.width, r0.height));
        
        r0 = A;
        r1 = C;
        assertTrue(Utils.isBoxesOverlapping(r0.x, r0.y, r0.width, r0.height, r1.x, r1.y, r1.width, r1.height));
        assertTrue(Utils.isBoxesOverlapping(r1.x, r1.y, r1.width, r1.height, r0.x, r0.y, r0.width, r0.height));

        r0 = B;
        r1 = C;
        assertTrue(Utils.isBoxesOverlapping(r0.x, r0.y, r0.width, r0.height, r1.x, r1.y, r1.width, r1.height));
        assertTrue(Utils.isBoxesOverlapping(r1.x, r1.y, r1.width, r1.height, r0.x, r0.y, r0.width, r0.height));

    }
    
    public void test_isBoxesOverlappingNeg() {
        Rectangle A = new Rectangle(0,0,1,1);
        Rectangle B = new Rectangle(1,1,2,2);
        Rectangle C = new Rectangle(100,100,1,1);
        
        Rectangle r0 = A;
        Rectangle r1 = B;       
        assertFalse(Utils.isBoxesOverlapping(r0.x, r0.y, r0.width, r0.height, r1.x, r1.y, r1.width, r1.height));
        assertFalse(Utils.isBoxesOverlapping(r1.x, r1.y, r1.width, r1.height, r0.x, r0.y, r0.width, r0.height));
        
        r0 = A;
        r1 = C;
        assertFalse(Utils.isBoxesOverlapping(r0.x, r0.y, r0.width, r0.height, r1.x, r1.y, r1.width, r1.height));
        assertFalse(Utils.isBoxesOverlapping(r1.x, r1.y, r1.width, r1.height, r0.x, r0.y, r0.width, r0.height));
        
        r0 = A;
        r1 = C;
        assertFalse(Utils.isBoxesOverlapping(r0.x, r0.y, r0.width, r0.height, r1.x, r1.y, r1.width, r1.height));
        assertFalse(Utils.isBoxesOverlapping(r1.x, r1.y, r1.width, r1.height, r0.x, r0.y, r0.width, r0.height));

        r0 = B;
        r1 = C;
        assertFalse(Utils.isBoxesOverlapping(r0.x, r0.y, r0.width, r0.height, r1.x, r1.y, r1.width, r1.height));
        assertFalse(Utils.isBoxesOverlapping(r1.x, r1.y, r1.width, r1.height, r0.x, r0.y, r0.width, r0.height));
        
        
    }
    
    public void testBubblesort() {
    	Integer [] sortable = {  4,   2,   1,   6,   2,   8,   0 };
    	String [] letters =   { "D", "C", "B", "E", "C", "G", "A" };
    	
    	Utils.bubbleSort(sortable, letters);
    	
    	System.out.println(Arrays.toString(sortable));
    	System.out.println(Arrays.toString(letters));
    }

    public void testWeightedRandomGet() {
        List<String> items = Arrays.asList("A", "B", "C");
        int [] weights = { 1, 0, 1 };

        for (int i=0; i<10000; i++)
            assertFalse(Utils.chooseRandomWeightedItem(items, weights).equals("B"));
    }

    public void testRotateArray() {

        Integer [] array = { 1,1,1,3,3,3 };
        Integer [] result = new Integer[3];
        Utils.rotate(array, result);
        Assert.assertTrue(Arrays.equals(result, new Integer[] { 1,1,1 }));
        Assert.assertTrue(Arrays.equals(array, new Integer[] { 3,3,3,1,1,1 }));

    }

    public void testColorParsing() throws Exception {

        assertEquals(GColor.fromString("[255,0,0,0]"), GColor.BLACK);
        assertEquals(GColor.fromString("ARGB[255,255,255,255]"), GColor.WHITE);

    }

    public void testRangeIter() {
        Utils.RangeIter iter = new Utils.RangeIter(0, 10);
        int idx = 0;
        while (iter.hasNext()) {
            int nxt = iter.next();
            assertEquals(nxt, idx++);
            System.out.print(" " + nxt);
        }
        assertTrue(idx == 11);

        System.out.println();
        idx = 10;
        iter = new Utils.RangeIter(10, 20, 2);
        while (iter.hasNext()) {
            int nxt = iter.next();
            assertEquals(nxt, idx);
            idx+=2;
            System.out.print(" " + nxt);
        }
        assertTrue(idx == 22);

        iter = new Utils.RangeIter(0, 0);
        assertTrue(iter.hasNext());
        assertEquals(new Integer(0), iter.next());

        iter = new Utils.RangeIter(0, -1);
        assertFalse(iter.hasNext());

    }

    public void testRomanNumeral() {
        RomanNumeral roman = new RomanNumeral();
        for (int i=0; i<=200; i++) {
            System.out.println(String.format("%-10d = %s", i, roman.toRoman(i)));
        }
    }

    public void testWrapText() {
        String txt = "\n\nHello\n\nGoodbye\n\n";
        String [] lines = Utils.wrapText(txt, 100);
        assertTrue(lines.length == 7);
        assertEquals("", lines[0]);
        assertEquals("", lines[1]);
        assertEquals("Hello", lines[2]);
        assertEquals("", lines[3]);
        assertEquals("Goodbye", lines[4]);
        assertEquals("", lines[5]);
        assertEquals("", lines[6]);
    }

    public void testContains() {

        GRectangle a = new GRectangle(-.5f, -.5f, 1, 1);
        GRectangle b = new GRectangle(-1, -1, 2, 2);

        assertTrue(b.contains(a));
        assertFalse(a.contains(b));

        System.out.println(b.getDeltaToContain(a));
        System.out.println(b.getDeltaToContain(a.movedBy(1, 1)));
        System.out.println(b.getDeltaToContain(a.movedBy(-1, 1)));
        System.out.println(b.getDeltaToContain(a.movedBy(-1, -1)));
        System.out.println(b.getDeltaToContain(a.movedBy(1, -1)));

    }

    public void testRectAddEq() {

        GRectangle a = new GRectangle(0, 2, 1, 6);
        System.out.println(a);
        a.setAspect(10f / 6);
        System.out.println(a);
    }


}

File: GameLibrary2/test/cc/lib/game/TestUtils.kt
package cc.lib.game

import cc.lib.game.GColor.Companion.fromString
import cc.lib.math.isBoxesOverlapping
import cc.lib.utils.bubbleSort
import cc.lib.utils.randomWeighted
import cc.lib.utils.rotate
import cc.lib.utils.wrap
import junit.framework.TestCase
import java.awt.Rectangle
import java.util.Arrays

class TestUtils : TestCase() {
	fun test_isBoxesOverlapping() {
		val A = Rectangle(0, 0, 1, 1)
		val B = Rectangle(-1, -1, 3, 3)
		val C = Rectangle(-2, 0, 5, 1)
		var r0 = A
		var r1 = B
		assertTrue(isBoxesOverlapping(r0.x, r0.y, r0.width, r0.height, r1.x, r1.y, r1.width, r1.height))
		assertTrue(isBoxesOverlapping(r1.x, r1.y, r1.width, r1.height, r0.x, r0.y, r0.width, r0.height))
		r0 = A
		r1 = A
		assertTrue(isBoxesOverlapping(r0.x, r0.y, r0.width, r0.height, r1.x, r1.y, r1.width, r1.height))
		assertTrue(isBoxesOverlapping(r1.x, r1.y, r1.width, r1.height, r0.x, r0.y, r0.width, r0.height))
		r0 = A
		r1 = C
		assertTrue(isBoxesOverlapping(r0.x, r0.y, r0.width, r0.height, r1.x, r1.y, r1.width, r1.height))
		assertTrue(isBoxesOverlapping(r1.x, r1.y, r1.width, r1.height, r0.x, r0.y, r0.width, r0.height))
		r0 = B
		r1 = C
		assertTrue(isBoxesOverlapping(r0.x, r0.y, r0.width, r0.height, r1.x, r1.y, r1.width, r1.height))
		assertTrue(isBoxesOverlapping(r1.x, r1.y, r1.width, r1.height, r0.x, r0.y, r0.width, r0.height))
	}

	fun test_isBoxesOverlappingNeg() {
		val A = Rectangle(0, 0, 1, 1)
		val B = Rectangle(1, 1, 2, 2)
		val C = Rectangle(100, 100, 1, 1)
		var r0 = A
		var r1 = B
		assertFalse(isBoxesOverlapping(r0.x, r0.y, r0.width, r0.height, r1.x, r1.y, r1.width, r1.height))
		assertFalse(isBoxesOverlapping(r1.x, r1.y, r1.width, r1.height, r0.x, r0.y, r0.width, r0.height))
		r0 = A
		r1 = C
		assertFalse(isBoxesOverlapping(r0.x, r0.y, r0.width, r0.height, r1.x, r1.y, r1.width, r1.height))
		assertFalse(isBoxesOverlapping(r1.x, r1.y, r1.width, r1.height, r0.x, r0.y, r0.width, r0.height))
		r0 = A
		r1 = C
		assertFalse(isBoxesOverlapping(r0.x, r0.y, r0.width, r0.height, r1.x, r1.y, r1.width, r1.height))
		assertFalse(isBoxesOverlapping(r1.x, r1.y, r1.width, r1.height, r0.x, r0.y, r0.width, r0.height))
		r0 = B
		r1 = C
		assertFalse(isBoxesOverlapping(r0.x, r0.y, r0.width, r0.height, r1.x, r1.y, r1.width, r1.height))
		assertFalse(isBoxesOverlapping(r1.x, r1.y, r1.width, r1.height, r0.x, r0.y, r0.width, r0.height))
	}

	fun testBubblesort() {
		val sortable = arrayOf(4, 2, 1, 6, 2, 8, 0)
		val letters = arrayOf("D", "C", "B", "E", "C", "G", "A")
		bubbleSort(sortable, letters)
		println(Arrays.toString(sortable))
		println(Arrays.toString(letters))
	}

	fun testWeightedRandomGet() {
		val items: List<String> = mutableListOf("A", "B", "C")
		val weights = intArrayOf(1, 0, 1)
		for (i in 0..9999)
			assertFalse(weights.randomWeighted(items).equals("B"))
	}

	fun testRotateArray() {
		val array = arrayOf(1, 2, 3, 4, 5, 6)

		assertEquals(array.rotate(1), arrayOf(2, 3, 4, 5, 6, 1))
	}

	@Throws(Exception::class)
	fun testColorParsing() {
		assertEquals(fromString("[255,0,0,0]"), GColor.BLACK)
		assertEquals(fromString("ARGB[255,255,255,255]"), GColor.WHITE)
	}


	fun testRomanNumeral() {
		for (i in 0..200) {
			println(String.format("%-10d = %s", i, RomanNumeral.toRoman(i)))
		}
	}

	fun testWrapText() {
		val txt = "\n\nHello\n\nGoodbye\n\n"
		val lines: Array<String> = txt.wrap(100)
		assertTrue(lines.size == 7)
		assertEquals("", lines[0])
		assertEquals("", lines[1])
		assertEquals("Hello", lines[2])
		assertEquals("", lines[3])
		assertEquals("Goodbye", lines[4])
		assertEquals("", lines[5])
		assertEquals("", lines[6])
	}

	fun testContains() {
		val a = GRectangle(-.5f, -.5f, 1f, 1f)
		val b = GRectangle(-1f, -1f, 2f, 2f)
		assertTrue(b.contains(a))
		assertFalse(a.contains(b))
		println(b.getDeltaToContain(a))
		println(b.getDeltaToContain(a.movedBy(1f, 1f)))
		println(b.getDeltaToContain(a.movedBy(-1f, 1f)))
		println(b.getDeltaToContain(a.movedBy(-1f, -1f)))
		println(b.getDeltaToContain(a.movedBy(1f, -1f)))
	}

	fun testRectAddEq() {
		val a = GRectangle(0f, 2f, 1f, 6f)
		println(a)
		a.setAspect(10f / 6)
		println(a)
	}
}
